### PR TITLE
feat: introduce Notification component

### PR DESCRIPTION
**Pull request by austin94**
#### Minor: New Feature

This PR introduces the `Notification` (in-line notification) component.  It includes 4 styles: `primary (default), success, warning, and error`.

screen shot 2018-11-29 at 3 17 17 pm
screen shot 2018-11-29 at 3 17 27 pm
screen shot 2018-11-29 at 3 17 36 pm
screen shot 2018-11-29 at 3 18 20 pm

#### Tests Added + Pass

Yes

#### License

MIT

### CODE REVIEW

**Comment by gergelyke** [comment]
do you think we can build this on top of the Toast component with style overrides?

**Comment by chasestarr** [comment]
Deploy preview for *baseui* ready!

Built with commit 0641690ea962381667ef81ed5951d8b67fd6b167

https://deploy-preview-566--baseui.netlify.com

**Comment by nadiia** [comment]
> do you think we can build this on top of the Toast component with style overrides?

That was the plan. Not build on top tho but add additional types, since the current toast component is an inline notification.

**Comment by austin94** [comment]
> do you think we can build this on top of the Toast component with style overrides?

I originally went down this path, but ran into some complexity.  With the amount of style overrides necessary it was easiest to just treat them as separate components. This also allows them to be more easily themed separately without Toast items cascading into Notification.

File: src/notification/examples.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    return (
+      <Notification kind={KIND.error}>
+        This is an error notification
+      </Notification>
+    );
+  },
+  [examples.OVERRIDES]: function Story5() {
+    return (
+      <Notification
+        overrides={{
```
Comment (blackswanny) [inline src/notification/examples.js]: we should be able to override animation of Notification. Can you add the test showing it?

**Comment by nadiia** [comment]
The only two style changes here based on the `kind` will be `backgroundColor` and `color`, and probably set the `closeable` default value to `false` unless this prop is set explicitly.

File: src/notification/notification.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import type {NotifcationPropsT} from './types';
+import {KIND} from './constants';
+import {StyledRoot} from './styled-components';
+import {getOverrides} from '../helpers/overrides';
+
+function Notification({children, kind, overrides = {}}: NotifcationPropsT) {
+  const [Root, baseRootProps] = getOverrides(overrides.Root, StyledRoot);
+
+  return (
+    <Root role="alert" $kind={kind} {...baseRootProps}>
```
Comment (blackswanny) [inline src/notification/notification.js]: they say that not all notifications should have `alert` role as it is considered more intense. Think for notification not error (or even warning) we shouldn't use this role

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_alert_role

**Comment by blackswanny** [comment]
how to close notification? Does it have timeout or optional closable button as Toast?

File: src/notification/examples.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+*/
+
+// @flow
+
+import React from 'react';
+import {Notification, KIND} from './';
+import examples from './examples-list';
+
+export default {
+  [examples.DEFAULT]: function Story1() {
```
Comment (blackswanny) [inline src/notification/examples.js]: any example of more complex content layout in notification?

**Comment by austin94** [comment]
> how to close notification? Does it have timeout or optional closable button as Toast?

These are in-inline notifications, not Toasts. Normally these are not able to be closed.

File: src/notification/styled-components.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+import {styled} from '../styles/index';
+import type {StyledRootPropsT} from './types';
+import {KIND} from './constants';
+
+export const StyledRoot = styled('div', ({$theme, $kind}: StyledRootPropsT) => {
+  let color = $theme.colors.notificationPrimaryText;
+  let backgroundColor = $theme.colors.notificationPrimaryBackground;
+
+  if ($kind === KIND.success) {
```
Comment (blackswanny) [inline src/notification/styled-components.js]: this seems to be a `switch...case` statement

File: src/notification/notification.js (hunk 2 above)
Comment (austin94) [reply src/notification/notification.js]: Maybe I'm miss-reading the docs, but `error` and `warning` are specifically shown using the role. Could you link to an example where a warning/error shouldn't have that role?

These should be used sparingly by design/dev, but if a notification is shown it should probably use the assertive tone to be read, right?

File: src/notification/types.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import type {OverrideT} from '../helpers/overrides';
+
+export type OverridesT = {
+  Root?: OverrideT<*>,
+};
+
+export type NotifcationPropsT = {
+  children?: Node,
+  overrides?: OverridesT,
+  kind: $Keys<typeof KIND>,
```
Comment (blackswanny) [inline src/notification/types.js]: $Values is more correct, but we need to freeze it https://github.com/uber-web/baseui/issues/539

File: src/notification/styled-components.js (hunk 4 above)
Comment (austin94) [reply src/notification/styled-components.js]: (I'm also used to an eslint rule that doesn't allow them 😄. This is now updated to a switch.

File: src/notification/types.js (hunk 5 above)
Comment (austin94) [reply src/notification/types.js]: We need to make this update in several areas. The majority of components are using `$Keys`.

Comment (austin94) [reply src/notification/types.js]: Updated and frozen.

**Comment by blackswanny** [comment]
> These are in-inline notifications, not Toasts. Normally these are not able to be closed.

Ok. Can you clarify with design that it's expected? Cause I anticipate that notifications can stack one by one taking space as I see them in some Uber apps. @nguyensomniac

File: src/notification/notification.js (hunk 2 above)
Comment (blackswanny) [reply src/notification/notification.js]: Error and warning's are good. But as per design we have `primary` and `success` notifications. They don't look like `assertive`

Comment (austin94) [reply src/notification/notification.js]: The `assertive` value is specifically a `POLITENESS_SETTING` and has not direction relation to what is actually being displayed.

It only means that the user will be immediately told of the notification, which should probably occur if a success notification is triggered.

Comment (blackswanny) [reply src/notification/notification.js]: ok, makes sense

**Comment by nguyensomniac** [comment]
> > These are in-inline notifications, not Toasts. Normally these are not able to be closed.
> 
> Ok. Can you clarify with design that it's expected? Cause I anticipate that notifications can stack one by one taking space as I see them in some Uber apps. @nguyensomniac

Yeah these aren't meant to be closed. In theory you could restyle a toast to look like an inline notification but you really shouldn't.

**Comment by austin94** [comment]
Closing in favor of #574

**Pull request closed without merging**